generators/add-licence-headers/index.js

```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  _generateComments() {
+    // get maximum line length by looping through all lines, returning the line length and passing them to Math.max
+    const lineArray = this.licenceText.split('\n');
+    const maxLineLength = Math.max(...lineArray.map((line) => line.length));
+
+    const align = (string, threshold, timesSpaces) => {
+      if (!timesSpaces) {
+        timesSpaces = threshold;
+      }
+
+      if (string.length > threshold) {
+        return ' '.repeat(timesSpaces);
+      }
+      return '';
+    };
+
+    Object.keys(comments).forEach((fileType) => {
+      const commentStyle = comments[fileType];
+      let comment = commentStyle.begin + commentStyle.body.repeat(maxLineLength) + '\n';
+      lineArray.forEach((line) => {
+        // only add a space character when the opening comment contains more than one charcters (e.g. to align the asterisks in .ts files vs. aligning the hashes in .py files)
+        comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+        comment += `${commentStyle.body}${align(line, 0, 1) + line}\n`;
+      });
+
+      comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+      comment += `${commentStyle.body.repeat(maxLineLength)}${commentStyle.end}`;
+      this.comments[fileType] = comment;
+    });
+  }
+
+  _getSourceFolders() {
+    const path = this.destinationPath().split('/');
+    return ['src', path[path.length - 1]];
+  }
+
+  _abort(msg) {
+    console.log('ABORTING');
+    return Promise.reject(msg ? msg : 'Step Failed: Aborting');
+  }
+
+  _addComments(path, fileExtension) {
+    glob(`**/*.${fileExtension}`, {cwd: path}, (err, files) => {
+      if (err) {
+        throw err;
+      }
+
+      if (files.length === 0) {
+        return;
+      }
+
+      files.forEach((file) => {
+        const filePath = path + '/' + file;
+        let fileContents = this.fs.read(filePath);
+
+        // TODO: override any comment if the file starts with one (e.g. when the licence changes)
+        // whenever a file starts with our licence header skip for now
+        if (fileContents.startsWith(this.comments[fileExtension])) {
+          return;
+        }
+
+        fileContents = this._findAndRemoveHeader(fileContents, fileExtension);
+
+        const newContents = this.comments[fileExtension] + '\n\n' + fileContents;
+        this.fs.write(filePath, newContents);
+      });
+    });
+  }
+
+  /**
+   * remove a multiline comment at the beginning of a file (e.g. licence header) if it exists
+   * @param fileContents
+   * @param fileExtension
+   * @returns {string}
+   * @private
+   */
+  _findAndRemoveHeader(fileContents, fileExtension) {
+    if (fileContents.startsWith(comments[fileExtension].begin)) {
+      const linesArray = fileContents.split('\n');
+      let line = linesArray[0];
+      while (line.startsWith(comments[fileExtension].begin) || line.indexOf(comments[fileExtension].body) > -1) {
+        line = linesArray.splice(0, 1)[0];
```

Comment:
splice returns the deleted line as far as I know. Thus wouldn't you check the first line but remove the second in the second iteration?

```
v = [a, b, c]
line = v[0]  === a
while (line.startsWith(a)) {
  removed = v.splice(0, 1); 
  v === [b, c] && removed === [a]
  line = removed[0] 
  -> line = a again
}
```

Reply:
right, I need to `let line = linesArray.splice(0, 1)[0];` before the while loop already, if I understand you correctly

Reply:
btw. `linesArray.splice(0, 1)[0]` <=> `linesArray.shift()`

Reply:
ah, right. good one! 👍